map_model/src/make/traffic_signals.rs

```diff
@@ -329,6 +332,426 @@ fn stage_per_road(map: &Map, i: IntersectionID) -> ControlTrafficSignal {
+                        if duration < MIN_CROSSWALK_TIME {
+                            duration = MIN_CROSSWALK_TIME;
+                        }
+                        s.stage_type = StageType::Variable(
+                            duration,
+                            Duration::const_seconds(1.0),
+                            Duration::const_seconds(1.0),
+                        )
+                    }
+                    _ => (),
+                }
+            }
+        }
+    }
+
+    fn merge_stages(ts: &ControlTrafficSignal) -> Option<ControlTrafficSignal> {
+        for s_src in &ts.stages {
+            // s_src is the stage we want to apply to the other stages
+            for s_dst in &ts.stages {
+                if s_src == s_dst {
+                    continue;
+                }
+                let mut merged_stage = s_dst.clone();
+                for m in &s_src.protected_movements {
+                    merged_stage.protected_movements.insert(m.clone());
+                }
+                let mut maybe_ts = ts.clone();
+                // insert at the head, keeping crosswalk last
+                maybe_ts.stages.insert(0, merged_stage);
+                if maybe_ts.validate().is_ok() {
+                    println!("can merge s_src={:#?} and s_dst={:#?}", s_src, s_dst);
+                    let mut stages: Vec<Stage> = Vec::new();
+                    for s in maybe_ts.stages {
+                        if s != *s_src && s != *s_dst {
+                            stages.push(s);
+                        }
+                    }
+                    maybe_ts.stages = stages;
+                    return Some(maybe_ts);
+                }
+            }
+        }
+        None
+    }
+
+    // sometimes protected oncoming left turns aren't possible.
+    fn is_conflict(ts: &ControlTrafficSignal, stage: &Stage) -> Option<(MovementID, MovementID)> {
+        for m1 in stage.protected_movements.iter().map(|m| &ts.movements[m]) {
+            for m2 in stage.protected_movements.iter().map(|m| &ts.movements[m]) {
+                if m1.conflicts_with(m2) {
+                    return Some((m1.id, m2.id));
+                }
+            }
+        }
+        return None;
+    }
+
+    fn protected_yield_stage(p: MovementID, y: MovementID) -> Stage {
+        let mut stage = Stage::new();
+        stage.protected_movements.insert(p);
+        stage.yield_movements.insert(y);
+        stage
+    }
+
+    fn multi_way_stages(map: &Map, id: IntersectionID) -> Option<ControlTrafficSignal> {
+        let mut ts = new(id, map);
+        let (mut right, mut left, straight, mut roads) = movements(&ts);
+        let (one_way, two_way) = straight_types(&straight);
+        for m in &one_way {
+            let mut stage = Stage::new();
+            stage.protected_movements.insert(m.clone());
+            for t in turns(&m.from.id, &right) {
+                stage.protected_movements.insert(t.clone());
```

Comment:
The `t.clone()` is redundant, you could just `insert(t)`

Reply:
I'm going to convert to turns().iter().for_each(|t|...); Let me know if you'd prefer the for loop instead.